File: CPP_Compiler/Analysis/Scan.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include "../Define/Token.h"
using namespace std;

static wstring::iterator current;

enum class CharType
{
    Unknown,
    WhiteSpace,
    NumberLiteral,
    StringLiteral,
    IdentifierAndKeyword,
    OperatorAndPunctuator
};

wstring scan_files(int fileCount, char const *files[])
{
    ifstream fin;
    string file;

    for (int i = 1; i < fileCount; i++)
    {
        int linesCount = 0;
        fin.open(files[i]);

        // 파일 여는데 실패했을 경우 처리
        if (!fin.is_open())
        {
            cerr << "Error opening file: " << files[i] << endl;
            return nullptr;
        }

        // file += '\n';
        string line;
        while (getline(fin, line))
        {
            file += line;
            file += "\n";
        }

        fin.close();
    }

    return wstring(file.begin(), file.end());
}

CharType get_char_type(wchar_t c)
{
    if (L' ' == c || L'\t' == c || L'\r' == c || L'\n' == c)
        return CharType::WhiteSpace;
    if (L'0' <= c && L'9' >= c)
        return CharType::NumberLiteral;
    if (L'\'' == c || L'\"' == c)
        return CharType::StringLiteral;
    if (L'가' <= c && L'힣' >= c)
        return CharType::IdentifierAndKeyword;
    return CharType::Unknown;
}

bool is_char_type(wchar_t c, CharType type)
{
    switch (type)
    {
    case CharType::NumberLiteral:
        return L'.' || L'0' <= c && L'9' >= c;
    case CharType::StringLiteral:
        return L'\'' != c && L'\"' != c;
    case CharType::IdentifierAndKeyword:
        return L'0' <= c && L'9' >= c ||
               L'가' <= c && L'힣' >= c;
    default:
        return false;
    }
}

Token scan_number_literal()
{
    wstring s;
    while (is_char_type(*current, CharType::NumberLiteral))
        s += *current++;
    return Token{Kind::NumberLiteral, s};
}

Token scan_string_literal()
{
    wstring s;
    current++;
    while (is_char_type(*current, CharType::StringLiteral))
        s += *current++;
    if (L'\'' != *current && L'\"' != *current)
    {
        cout << "문자열이 종료되지 않았습니다." << endl;
        exit(1);
    }
    return Token{Kind::StringLiteral, s};
}

Token scan_identifier_and_Keyword()
{
    wstring s;
    while (is_char_type(*current, CharType::StringLiteral))
        s += *current++;
    
}

vector<Token> scan_token(int fileCount, char const *files[])
{
    wstring code = scan_files(fileCount, files);
    vector<Token> result;
    code += L'\0';
    current = code.begin();
    while (*current != L'\0')
    {
        // wprintf(L"%c",*current++);
        switch (get_char_type(*current))
        {
        case CharType::WhiteSpace:
            current++;
            break;
        case CharType::NumberLiteral:
            result.push_back(scan_number_literal());
            break;
        case CharType::StringLiteral:
            result.push_back(scan_string_literal());
            break;
        case CharType::IdentifierAndKeyword:
            result.push_back(scan_identifier_and_Keyword());
        default:
            break;
        }
    }
    result.push_back({Kind::EndOfCode});
    return result;
}

File: Rust_Compiler/src/analysis/scan.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;
use define::kind::Token;
use define::token::Kind;

enum CharType {
    Unknown,
    WhiteSpace,
    NumberLiteral,
    StringLiteral,
    IdentifierAndKeyword,
    OperatorAndPunctuator,
}

fn get_char_type(c: char) -> CharType {
    match c {
        ' ' | '\t' | '\r' | '\n' => CharType::WhiteSpace,
        '0'..='9' => CharType::NumberLiteral,
        '\'' | '\"' => CharType::StringLiteral,
        '가'..='힣' => CharType::IdentifierAndKeyword,
        _ => CharType::Unknown,
    }
}

fn is_char_type(c: char, kind: CharType) -> bool {
    match kind {
        CharType::NumberLiteral => c == '.' || ('0'..='9').contains(&c),
        CharType::StringLiteral => c != '\'' && c != '\"',
        CharType::IdentifierAndKeyword => ('0'..='9').contains(&c) || ('가'..='힣').contains(&c),
        _ => false,
    }
}

fn scan_number_literal(input: &mut std::str::Chars) -> Token {
    let mut value = String::new();
    while let Some(c) = input.next() {
        if is_char_type(c, CharType::NumberLiteral) {
            value.push(c);
        } else {
            break;
        }
    }
    Token::new(Kind::NumberLiteral, value)
}

fn scan_string_literal(input: &mut std::str::Chars) -> Token {
    let mut value = String::new();
    input.next(); // Skip the opening quote
    while let Some(c) = input.next() {
        if is_char_type(c, CharType::StringLiteral) {
            value.push(c);
        } else if c != '\'' && c != '\"' {
            println!("문자열이 종료되지 않았습니다.");
            std::process::exit(1);
        } else {
            break;
        }
    }
    Token::new(Kind::StringLiteral, value)
}

fn scan_identifier_and_keyword(input: &mut std::str::Chars) -> Token {
    let mut value = String::new();
    while let Some(c) = input.next() {
        if is_char_type(c, CharType::IdentifierAndKeyword) {
            value.push(c);
        } else {
            break;
        }
    }
    // Determine if the identifier is a keyword
    let kind = match value.as_str() {
        // Add more keyword patterns here if needed
        "if" => Kind::If,
        "else" => Kind::Else,
        _ => Kind::Identifier,
    };
    Token::new(kind, value)
}

fn scan_files(file_paths: Vec<&str>) -> String {
    let mut file_content = String::new();
    for file_path in file_paths {
        let path = Path::new(file_path);
        let file = match File::open(&path) {
            Ok(file) => file,
            Err(_) => {
                eprintln!("Error opening file: {}", file_path);
                return String::new();
            }
        };
        let reader = BufReader::new(file);
        for line in reader.lines() {
            if let Ok(line) = line {
                file_content.push_str(&line);
                file_content.push('\n');
            }
        }
    }
    file_content
}

fn scan_tokens(file_paths: Vec<&str>) -> Vec<Token> {
    let file_content = scan_files(file_paths);
    let mut tokens = Vec::new();
    let mut input = file_content.chars().peekable();

    while let Some(&c) = input.peek() {
        match get_char_type(c) {
            CharType::WhiteSpace => {
                input.next();
            }
            CharType::NumberLiteral => {
                tokens.push(scan_number_literal(&mut input));
            }
            CharType::StringLiteral => {
                tokens.push(scan_string_literal(&mut input));
            }
            CharType::IdentifierAndKeyword => {
                tokens.push(scan_identifier_and_keyword(&mut input));
            }
            CharType::Unknown => {
                input.next();
            }
            _ => {}
        }
    }

    tokens.push(Token::new(Kind::EndOfCode, String::new()));
    tokens
}

fn main() {
    let file_paths = vec!["path/to/your/file1.txt", "path/to/your/file2.txt"];
    let tokens = scan_tokens(file_paths);
    println!("{:#?}", tokens);
}
